Sprint 1: increment 4

bq2cd opened the pull request:
This is the final increment of the first sprint.

It contains command-line arguments parsing logic
for agent and server binaries.
There are also changes to the agent internal architecture
(e.g. reducing number of components from 3 to 2) and general
refactoring to support better testability.

High-level description of the changes in the previous
increments can be found in the corresponding pull requests:
- #3
- #2 
- #1

bq2cd committed:
Module initialisation + testing dependencies

'testify' will be used as a main library when writing tests.

bq2cd committed:
Ignore extra local files

bq2cd committed:
Basic skeleton: repository, service, handler

Public interfaces for repository and service packages:
- repository.Storage
- service.Metrics

Private implementation skeleton for the public interfaces:
- repository.memStorage
- service.metricService

Main HTTP handler with routing support (via http.ServeMux):
- handler.Router

Basic config struct for server.
- server.Config

Server statup process defined.

bq2cd committed:
Move NewMetricFromURLPath: model -> handler

It is more sensible to place NewMetricFromURLPath function
into 'handler' package since it deals with URLs.

bq2cd committed:
NewMetricFromURLPath implementation + tests

There is a concern about 'Hash' field being public and therefore
inadvertently modifiable by external actors (e.g. another internal package).
The problem is that we need to render this field to JSON and in the future
store it in the database as well, so it needs to be public for
'json.Marshal' to discover it (but that's only if the call to
'json.Marshal' happens in another package).
Alternatively, we could implement 'json.Marshaller' interface to serialize
private (non-exported) fields, or provide a method for a 'Metric' model
that would handle serialization (e.g. 'ToJSON()').
This would work for the JSON encoding.
For other serialisation protocols, another solution might be needed.

bq2cd committed:
memStorage implementation + tests

Simple in-memory map should suffice for the moment
(as per specification).

bq2cd committed:
service.Metrics implementation + tests

bq2cd committed:
handler tests

These are failing for now.
In the spirit of TDD, we're committing tests first.
The implementation will follow in a separate commit.

bq2cd committed:
handler implementation

Make tests green again.

'go fmt', 'goimports' and 'go vet' have been applied too.

bq2cd committed:
service: extend Metrics interface with Retrieve method

Also rename 'Update' -> 'Store' and allow passing multiple
metrics in one go (one metric is a must, optionally can pass more)
for caller's convenience.

'Retrieve' method will be useful down the line when we will need to
implement an API for metric retrieval, and also in agent, where we
could reuse Metrics service to store intermediate values before
reporting them to the server.

bq2cd committed:
refactor: use MetricKey instead of Hash in memStorage

Current commit introduces a predicable metric key based on its type and id
to be used internally in memStorage map (and possibly other places).
This replaces previously fragile approach with using Hash field as a key
when hash value was generated simply by combining metric type and id in a
custom string format; that approach could potentially lead to subtle bugs
when Hash field would inadvertently be updated in some place
to a different value, thus rendering Metric struct internally inconsistent.

Therefore, the metric can be reliable identified and addressed by its
MetricKey, while Hash field could be used for other purposes, e.g.
for verifying metric integrity.

bq2cd committed:
Avoid storing empty/invalid metrics

Metric is considered empty/invalid if any of the following conditions is met:
- empty id
- empty type
- type is counter but delta is nil
- type is gauge but value is nil
- type is custom but both value and delta are nil

Such metrics are silently dropped when attempting to store them via
service.Metrics
Internally, memStorage avoid storing emtpy metrics as well.

bq2cd committed:
Support returning all metrics from the storage

This will be useful when we would need to implement a functionality
that displays all metrics available at the server.
Also, this would be useful in the agent for temporarily storing metrics
before reporting them to an upstream.

bq2cd committed:
agent: basic interfaces + tests (failing as expected)

Here we define basic interfaces and structs with dummy methods that
will be implemented in the next commits.
This commit also defines tests which are failing at the moment;
this is expected (in the spirit of TDD).
Next commits will be focused on making the tests green again
(and potentially fixing/expanding the tests too).

bq2cd committed:
agent: cmd entrypoint with graceful shutdown support

By using 'signal.NotifyContext' we can propagate cancellation caused by
a set of signals (SIGINT, SIGTERM) to the internal loop and allow it
to terminate gracefully.

bq2cd committed:
agent: main loop implementation

Collecting and storing metrics is done synchronously as those operations
are expected to finish quickly - at least for now: the metrics we are
collecting are readily available to the agent process and do not involve
any I/O or network calls.

Reporting, however, is performed asynchronously, in a separate goroutine,
because there is network call involved and there might be timeouts, etc.
When agent context is cancelled (e.g by a SIGINT), main loop will wait
for all still running goroutines to finish and report their errors,
then it will stop the execution.

To ensure proper termination of reporting goroutines, we should pass
proper context to the reporter itself and ensure that it uses
this context for HTTP requests,
thus ensuring their cancellation should the context be cancelled.

bq2cd committed:
agent: reporter implementation + tests adjustments

Ensure that HTTP client gets proper timeout set on initialisation
of the agent process.
Also, ensure that each HTTP request has a context set, so that we can
cancel in-flight requests gracefully.

bq2cd committed:
agent: collect runtime metrics from runtime.MemStats

It was chosed to use reflection to retrieve actual values from the
'runtime.MemStat' struct to avoid duplicating metric names in multiple
places.

bq2cd committed:
agent: improve tests coverage

Poor test coverage for collector needs to further improved by
abstracting metrics source into an interface.

bq2cd committed:
agent: metrics source interface + tests

To add more flexibility to our metrics collector, it was decided to
abstract metrics source with an interface and support multiple sources
in the collector itself.

This commit introduces 'Source' interface and tests covering
dummy implementation.
The tests are failing as expected - this is to be fixed in the next
commits.

bq2cd committed:
agent: improved test coverage for metric sources

A few more abstractions had to be introduced to achieve better
testability, but in the end, test coverage significantly improved.

Also, this commit includes reference implementations that pass the tests.

bq2cd committed:
agent: refactor collector to use metric sources

This makes our default collector easier to extend in the future:
instead of implementing another collector, we would implement
another metric source.

bq2cd committed:
agent: improve test coverage

bq2cd committed:
agent: tweak main loop test

It appears that the test for slow reporter is a bit flaky
when it is running in Github action environment.
Increasing timeouts/intervals a bit to mitigate that.

bq2cd committed:
A package to convert url path to/from metric: stage 1

Converting a metric to an url path and vice versa is becoming an
increasingly common task as we are about to introduce
an API to read metrics from server.

This commit begins extraction of the necessary functionality to
'handler/urlpath' internal package, which will be conducted in 3 stages:
- stage 1 (this commit): interfaces + prototypes + tests;
- stage 2: implementation of the new package;
- stage 3: refactoring to move existing functionality into
  the new package;

Stay tuned :)

bq2cd committed:
A package to convert url path to/from metric: stage 2

Implementation of 'handler/urlpath' to match tests.

bq2cd committed:
A package to convert url path to/from metric: stage 3

Switched to the 'handler/urlpath' package.
A couple of tests required adjusting due to minor changes
in the way we parse url path now.

bq2cd committed:
use chi.Router instead of http.ServeMux

chi.Router is more flexible when it comes to defining routes, and it
also supports middlewares, which is quite handy.
Here we already enable logging middleware to help with the debugging.

bq2cd committed:
improve router tests

Now using proper HTTP server (from httptest) and sample HTTP requests
to validate that correct handlers are responding.
This overlaps a bit with tests for a concrete handler, but only
covers a happy path here as the goal is to demonstrate that
correct handler is picked by the router,
while in the tests for the handler we cover more edge cases.

bq2cd committed:
handlers for reading metrics: '/' and '/value': tests

Adding tests for the following future handlers:
- 'readHandler': responds on '/' with the list of all known metrics
- 'valueHandler': responds on '/value/*' with the value for a given metric

The implementation to follow in the next commit.

bq2cd committed:
handlers for reading metrics: '/' and '/value': implementation

Implementation for 'readHandler' and 'valueHandler' to conform with
the tests defined in the previous commit.

bq2cd committed:
improve test coverage

Added tests for edge cases.

bq2cd committed:
cmd: refactor 'main' function into smaller ones

No functionality change, only split 'main' function into smaller
functions with specific emphasis on isolating future arguments parsing
logic.

bq2cd committed:
model: introduce metric.Copy method

There is a subtle bug in the 'service.Metrics' - whenever we change a
delta for a counter to include previously stored value, the original
metric object also changes because it contains pointers to the actual
values.
To work around that, we need to copy original metric object before
changing its value.
Tests confirming the bug are also added.

bq2cd committed:
agent: serialize reporter calls

If reporter is being slow for some reason, we do not want to spawn
too many goroutines trying to send probably the same values to the
server - only goroutine should do the reporting, and when it finishes,
the next can proceed.
While a proper way to addressing this would be running reporter loop in
a separate "worker" goroutine, it would require too many changes in the
current "architecture", so we chose a simpler approach which requires
minimal changes: add a mutex and only proceed with reporting if mutex
is grabbed, otherwise new goroutine would return immediately.
Necessary tests are included to verify the solution.

bq2cd committed:
agent: storer should store only last value for counters

Given that we always add an incoming value for a counter to its previous
value on the server side, it does not make sense to do the same on the
agent side, because it would lead to inflated counter values on the
server.
For example, if we take 'PollCount' metric (which is the only counter
being reported for now), in the previous implementation the agent would
store values as 1, 3, 6, 10, ... at each poll interval, and the server
would store them as 1, 4, 10, 20 ...

To be fair, the current implementation of the storer gives little
advantage over using 'repository.Storage' directly and might seem an
unnecessary abstraction, still, we shall keep it for now in case
future requirements might justify it.
If that would not be the case, we will dismantle it at a later stage.

There is another potential problem with reporting counters to the server.
Even if we report the last value of the counter from the agent,
the server would still add this value to the previously stored one -
this might or might not be incorrect from the logical point of view.
One way to tackle that would be making reporter store the last value sent
to the server and only send deltas on subsequent reports
(only for counters, that is).
Still, on the first start of the agent, the first value reported would
again be incorrect if the server already had a previous value stored -
the agent would report the current value of the counter, and the server
would add that to the existing one.
Although, this would not be a problem for a 'PollCount' as it would
start from '0', but if it was a system counter that was never reset to
'0', and agent was restarted, the initial reported value would be
the last value of that counter, not the delta.
To mitigate that, we can either use a persistent storage for the reporter,
or retrieve the last known value from the server itself upon the first
report.

bq2cd committed:
agent: report only delta for counters

Here we attempt to make reporting of counters more accurate by reporting
only delta values instead of full values.

Why this matters?
Previously, the agent would store the last value of the counter and
report is as is to the server.
The server would add received value to the previously stored value, thus
inflating the real value of the counter.
For example, for 'PollCount' metric (the only counter for now), the
agent would report the following sequence
(assuming poll interval == report interval): 1, 2, 3, 4, 5, ...
The server would store another sequence: 1, 3, 6, 10, 15, ...
This does not seem correct from the logical point of view: the actual
number of polls by the agent in the example above would be 5,
but the server will store 15, not 5, which does not make sense.

To mitigate that, we use an internal storage (currently in-memory)
in the reporter itself, where it stores successfully sent metrics as is.
Whenever next round of reporting comes, the reporter would calculate
a delta for a given counter metric, and send the delta instead of the
full value.
This approach is not without issues, though.
For example, with current in-memory storage, whenever the agent is
restarted, it would send the full value on the first report.
If storage is not in-memory and cannot store the last value after
being successfully sent to the server, the next time the reporter
will report incorrect delta (or even full value).
A better approach might be having the reporter ask the server for the
last value on the first report and use it as a base to calculate a delta.
That said, we will stick with the current imperfect implementation
for the time being until further requirements emerge.

bq2cd committed:
agent: dispose of unnecessary Storer intermediary

When we introduced an internal storage to the reporter, the role of
'Storer' component in the agent pipeline had diminished and become
unnecessary (as it was simply a proxy to the underlying storage).
Therefore, it appeared to make more sense to get rid of this component
in favor of having internal storage in the collector itself,
where it could store intermediate results of the polling.
The implementation of that required changing 'Collector' interface
to introduce 'Snapshot()' method, which would return all metrics from
the collector's internal storage (similar to what 'Storer.Retrieve()' did).

bq2cd committed:
cmd: tests for arguments parsing logic

bq2cd committed:
cmd: implementation of arguments parsing logic

bq2cd committed:
server: graceful shutdown on SIGINT/SIGTERM

In this commit we implement graceful shutdown for the server.
To facilitate testing, the logic of running the server was refactored
out into 'internal/server' package and appropriate tests were added
on top of that.

bq2cd committed:
cmd: improve test coverage for 'server'/'agent' binaries

With a bit of refactoring, we're now able to test 'run' entrypoints
that launch a server and an agent.
As a result, 'main' function is very slim and only calls 'run' entrypoint
and deals with returned error.

bq2cd committed:
cmd: skip long-running tests in Github actions

Also, suppress stderr on flag parsing errors in tests.

bq2cd committed:
agent: decrease test sensitivity for 'Run' entrypoint

There might not be enough time/resources to complete the test with
a millisecond accuracy in the Github actions environment.
On the local machine, however, the test completes just fine.
Adding a few more milliseconds between poll interval and context timeout
should help.

bq2cd committed:
cmd: fix data races in agent tests

Perederey commented on internal/agent/agent.go:
```diff
@@ -0,0 +1,79 @@
+
+// Run launches main loop of the agent worker:
+// collecting metrics and reporting them back to a server.
+func (a *agentWorker) Run() error {
```
Метод `Run()` содержит слишком много логики. Стоит вынести обработку тикеров и горутин в отдельные методы

Perederey commented on internal/service/metrics.go:
```diff
@@ -0,0 +1,92 @@
+	"github.com/bq2cd/yp-go-metrics/internal/repository"
+)
+
+// Metrics defines an interface to work with metrics.
```
Интерфейс `Metrics` хорошо определен, но метод `Store` принимает ... параметры, что может усложнить использование - посмотри в сторону `StoreBatch`

Perederey commented on internal/agent/collector.go:
```diff
@@ -0,0 +1,56 @@
+	Snapshot() ([]model.Metric, error)
+}
+
+type defaultCollector struct {
```
Структура `defaultCollector` имеет хорошую инкапсуляцию, но зависимость от конкретного типа `repository.Storage` нарушает принцип инверсии зависимостей - надо использовать интерфейс

bq2cd replied:
`repository.Storage` и есть интерфейс, но переделал конструктор чтобы принимать интерфейс явно

Perederey commented on cmd/agent/main.go:
```diff
@@ -1,3 +1,122 @@
+	return ag.Run()
+}
+
+func parseArgs(fs *flag.FlagSet, args []string) (config.Config, error) {
```
Функция `parseArgs` слишком большая и выполняет много задач. Стоит разбить ее на более мелкие для читаемости

Perederey commented on internal/agent/reporter.go:
```diff
@@ -0,0 +1,85 @@
+	return &defaultReporter{context: ctx, client: client, reported: storage}
+}
+
+func (r *defaultReporter) reportSingle(metric model.Metric) error {
```
Метод `reportSingle` делает слишком много вещей. Стоит разделить логику на несколько методов

Perederey reviewed (commented):
Привет! Отлично справился с заданием! Реализована полноценная система сбора метрик с агентом и сервером, хорошая архитектура с разделением на слои, отличное покрытие тестами.

Из того, на что стоит обратить внимание:
- Некоторые функции слишком большие и делают много вещей - стоит разбивать их на более мелкие
- Есть места где нарушается принцип единственной ответственности
- Можно улучшить именование некоторых структур (убрать избыточность типа `agentWorker`)

А так молодец! Код работает, архитектура продуманная, тесты покрывают основную функциональность!

bq2cd committed:
service: refine Metrics interface to segregate single/batch operations

(based on Code Review comments)

Instead of using a single method to store either single or multiple
metrics depending on the number of arguments passed, it was decided
to define separate methods for each type of operation.
This way, when we want a single metric, we would call `StoreSingle`
or `RetrieveSingle` method and get either a metric or error in return.
For multiple metrics, we would use `StoreBatch`/`RetrieveBatch` methods
explicitly as we would need to handle returned values differently.
E.g. `RetrieveBatch` would only return existing metrics without
throwing an error by silently skipping non-existent keys.
Finally, it was decided to keep `RetrieveAll` as a separate method
instead of generalising `RetrieveBatch` to return all metrics
if no specific keys were requested - this way a caller would express
a specific intent.

PS. We could've kept original `Store`/`Retrieve` names for methods
operating on a single metric/key, but we opted to provide more explicit
naming instead.

bq2cd committed:
agent: decompose 'Run' entrypoint

Extract a generic function that can run a periodic task at a given interval
and optionally with a custom initial delay.
Now, 'Run' function can launch 2 goroutines (collector and reporter)
that utilise that periodic function and waits for them to finish
before exiting.
Graceful shutdown has moved from the 'Run' function into the
periodic function which watches for context cancellations and terminates
when the last chunk of work has completed.

bq2cd committed:
reporter: decompose 'reportSingle' method

There are 3 independent parts in that function: getting adjusted
(sendable) metric value (as we only send deltas to maintain correct
values on the server side), sending metric to the server, and storing
metric that was successfully sent, to the internal storage.
Out of these parts, the first two are easily extractable into separate
methods, while the third part is a one-liner, so it stays inside
'reportSingle' function.
To be fair, there is quire a bit of copying going on for the 'Metric'
struct, so at some point we might need to get a pprof and switch to using
pointers if the pprof data hints at that. Or change internal
representation of the metric to something more performant.

bq2cd committed:
server: rename 'serverWorker' -> 'server'

There seems to be no need for extra suffix here.

bq2cd committed:
repository: ensure mem storage is concurrency-safe

Protect internal map with RWMutex to avoid data races and also segregate
read and write locks.
Since we don't necessarily have a heavy read load, the use of `sync.Map`
is not warranted in this case.

bq2cd committed:
cmd: better config population in 'parseArgs'

Here we provide a better (or more streamlined) way to set config
options by using so called "funcopts" approach.
That way we can encapsulate value validation into the option function
itself, and also have a high level validation function to check config
for logical errors.

bq2cd committed:
tests: add missing tests and improve coverage

bq2cd committed:
agent: remove unnecessary 'default' prefix

There isn't much sense in having `defaultCollector` (or `defaultReporter`)
instead of just `collector` - there isn't another collector anyway (yet).
So, in spirit of YAGNI and KISS let's keep shorter names.
